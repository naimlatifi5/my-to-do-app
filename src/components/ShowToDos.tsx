import React from 'react'
import { ToDo } from '../utils/model'
import ToDoItem from './ToDoItem'
import styled from 'styled-components'

interface Props {
  allToDos: ToDo[]
  setAllToDos: React.Dispatch<React.SetStateAction<ToDo[]>>
}
const List = styled.ul`
  padding: 0;
  margin: 0 0 48px 0;
  list-style: none;

  > li {
    em {
      font-size: 16px;
    }
    margin-bottom: 12px;
    display: flex;
    align-items: center;
    font-size: 18px;
    .todo-item {
      color: #3b3b71;
    }
    .done-item {
      font-style: italic;
      color: #8a8aaa;
    }
  }
`

const ShowToDos: React.FC<Props> = ({ allToDos, setAllToDos }) => {
  console.log(allToDos)
  let completedToDos = allToDos.filter((item) => item.isDone === true)
  let notCompletedToDos = allToDos.filter((item) => item.isDone === false)

  return (
    <>
      <List>
        {notCompletedToDos.length ? (
          notCompletedToDos.map((item) => {
            return (
              <ToDoItem
                key={item.id}
                item={item}
                allToDos={allToDos}
                setAllToDos={setAllToDos}
              />
            )
          })
        ) : (
          <li>
            <em data-testid="no-item">OBS! No item added yet...</em>
          </li>
        )}
      </List>
      {completedToDos.length ? (
        <h4>COMPLETED ({completedToDos.length})</h4>
      ) : null}
      <List>
        {completedToDos.map((item) => {
          return (
            <ToDoItem
              data-testid="completed-item"
              key={item.id}
              item={item}
              allToDos={allToDos}
              setAllToDos={setAllToDos}
            />
          )
        })}
      </List>
    </>
  )
}

export default ShowToDos
