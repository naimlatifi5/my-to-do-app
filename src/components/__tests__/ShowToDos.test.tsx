import {
  render,
  screen,
  waitFor,
  cleanup,
  fireEvent,
  getAllByTestId,
  within,
} from '@testing-library/react'
import ShowToDos from '../ShowToDos'

describe('Test suit for ShowToDos component', () => {
  const setAllToDos = jest.fn()
  it('It should render the component onto the screen', () => {
    expect(true).toBeTruthy()
  })

  it('It should render default text if there are no toDos added', async () => {
    const fakeAllToDos = []
    render(<ShowToDos allToDos={fakeAllToDos} setAllToDos={setAllToDos} />)
    expect(screen.getByTestId('no-item')).toHaveTextContent(
      'OBS! No item added yet...',
    )
  })

  it('It should render uncompleted todos in a list as snapshot', async () => {
    const fakeAllToDos = [{ id: 1659648862224, isDone: false, todo: "asdf'" }]
    const { asFragment } = render(
      <ShowToDos allToDos={fakeAllToDos} setAllToDos={setAllToDos} />,
    )
    // test to add something to the dom ul element inside component and you will see the component will fail because of the previsous snapshoot differs
    expect(asFragment()).toMatchSnapshot()
  })

  it('It should render uncompleted todos in a list and show the length', async () => {
    const mockedUnCompletedToDos = [
      { id: 1659648862224, isDone: false, todo: "asdf'" },
    ]
    render(
      <ShowToDos allToDos={mockedUnCompletedToDos} setAllToDos={setAllToDos} />,
    )
    const unCompletedToDos = screen
      .getAllByTestId('item')
      .map((item) => item.textContent)
    const fakeTodos = mockedUnCompletedToDos.map((c) => c.todo)
    expect(unCompletedToDos).toEqual(fakeTodos)
    expect(unCompletedToDos).toHaveLength(1)
  })
  it('It should render heading for completed todos and the list with the length', async () => {
    const mockedCompletedToDos = [
      { id: 1659648862224, isDone: true, todo: "completed'" },
    ]
    render(
      <ShowToDos allToDos={mockedCompletedToDos} setAllToDos={setAllToDos} />,
    )
    const completedToDos = screen
      .getAllByTestId('item')
      .map((item) => item.textContent)
    const fakeTodos = mockedCompletedToDos.map((c) => c.todo)
    expect(completedToDos).toEqual(fakeTodos)
    expect(completedToDos).toHaveLength(1)
    expect(screen.getByTestId('heading-completed')).toHaveTextContent(
      'COMPLETED (1)',
    )
  })
})
