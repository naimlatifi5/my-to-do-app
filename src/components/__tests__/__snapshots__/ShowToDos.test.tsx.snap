// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suit for ShowToDos component It should render uncompleted todos in a list as snapshot 1`] = `
<DocumentFragment>
  <ul
    class="sc-eCYdqJ bwEEBQ"
  >
    <li
      data-testid="item"
    >
      <span
        class="sc-bczRLJ cfrFOm"
      />
      <span
        class="todo-item"
      >
        asdf'
      </span>
      <span
        class="sc-dkzDqf kQnhMH"
      >
        <span
          class="sc-hKMtZM gvuFEE"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
            />
            <path
              d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
            />
          </svg>
        </span>
        <span
          class="sc-hKMtZM gvuFEE"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M864 256H736v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zm-200 0H360v-72h304v72z"
            />
          </svg>
        </span>
      </span>
    </li>
  </ul>
  <ul
    class="sc-eCYdqJ bwEEBQ"
  />
</DocumentFragment>
`;
